params/extras/config.go

```diff
@@ -0,0 +1,352 @@
+		rew = newtime
+	case newtime == nil || *storedtime < *newtime:
+		rew = storedtime
+	default:
+		rew = newtime
+	}
+	err := &ConfigCompatError{
+		What:         what,
+		StoredTime:   storedtime,
+		NewTime:      newtime,
+		RewindToTime: 0,
+	}
+	if rew != nil && *rew > 0 {
+		err.RewindToTime = *rew - 1
+	}
+	return err
+}
+
+// UnmarshalJSON parses the JSON-encoded data and stores the result in the
+// object pointed to by c.
+// This is a custom unmarshaler to handle the Precompiles field.
+// Precompiles was presented as an inline object in the JSON.
+// This custom unmarshaler ensures backwards compatibility with the old format.
+func (c *ChainConfig) UnmarshalJSON(data []byte) error {
+	// Alias ChainConfigExtra to avoid recursion
+	type _ChainConfigExtra ChainConfig
+	tmp := _ChainConfigExtra{}
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+
+	// At this point we have populated all fields except PrecompileUpgrade
+	*c = ChainConfig(tmp)
+
+	// Unmarshal inlined PrecompileUpgrade
+	return json.Unmarshal(data, &c.GenesisPrecompiles)
+}
+
+// MarshalJSON returns the JSON encoding of c.
+// This is a custom marshaler to handle the Precompiles field.
+func (c *ChainConfig) MarshalJSON() ([]byte, error) {
+	// Alias ChainConfigExtra to avoid recursion
+	type _ChainConfigExtra ChainConfig
+	tmp, err := json.Marshal(_ChainConfigExtra(*c))
+	if err != nil {
+		return nil, err
+	}
+
+	// To include PrecompileUpgrades, we unmarshal the json representing c
+	// then directly add the corresponding keys to the json.
+	raw := make(map[string]json.RawMessage)
+	if err := json.Unmarshal(tmp, &raw); err != nil {
+		return nil, err
+	}
+
+	for key, value := range c.GenesisPrecompiles {
+		conf, err := json.Marshal(value)
+		if err != nil {
+			return nil, err
+		}
+		raw[key] = conf
+	}
+
+	return json.Marshal(raw)
+}
+
+type fork struct {
+	name      string
+	block     *big.Int // some go-ethereum forks use block numbers
+	timestamp *uint64  // Avalanche forks use timestamps
+	optional  bool     // if true, the fork may be nil and next fork is still allowed
+}
+
+func (c *ChainConfig) CheckConfigForkOrder() error {
+	if c == nil {
+		return nil
+	}
+	// Note: In Avalanche, upgrades must take place via block timestamps instead
+	// of block numbers since blocks are produced asynchronously. Therefore, we do
+	// not check block timestamp forks in the same way as block number forks since
+	// it would not be a meaningful comparison. Instead, we only check that the
+	// Avalanche upgrades are enabled in order.
+	// Note: we do not add the precompile configs here because they are optional
+	// and independent, i.e. the order in which they are enabled does not impact
+	// the correctness of the chain config.
+	return checkForks(c.forkOrder(), false)
+}
+
+// checkForks checks that forks are enabled in order and returns an error if not.
+// [blockFork] is true if the fork is a block number fork, false if it is a timestamp fork
```

Comment:
nit brackets don't work for function arguments

```suggestion
// `blockFork` is true if the fork is a block number fork, false if it is a timestamp fork
```